File: Controller/PiCom/Clients/LANClient.py
"""
Example Client to be implemented by android and the pi
"""

import socket

from PiCom.Payload import PayloadEvent, PayloadType, Payload, send_payload, receive_payload

TIMEOUT = None


class LANClientHandler:
    def received(self, req_payload: Payload, res_payload: Payload):
        pass


class LANClient:
    def __init__(self, host: str, port: int, handler: LANClientHandler = None):
        self.HANDLER = handler
        self.HOST = host
        self.PORT = port
        self.sock = None
        self.open_connection()

    def close_connection(self):
        if not self.is_open():
            print("[!] Already Closed")
            return
        send_payload(self.sock, Payload("Goodbye",
                                        PayloadEvent.SYSTEM,
                                        PayloadType.END))
        self.sock.close()

    def open_connection(self):
        if self.is_open():
            print("[!] Already Open!")
            return
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.sock.settimeout(TIMEOUT)
            self.sock.connect((self.HOST, self.PORT))
        except Exception:
            self.sock.close()

    def send(self, payloads):
        if not self.is_open():
            self.open_connection()

        if isinstance(payloads, list):

            for payload in payloads:
                send_payload(self.sock, payload)
                if self.HANDLER is None:
                    raise SyntaxError("When sending multiple payloads, the handler must be specified.")
                self.HANDLER.received(self, payload, receive_payload(self.sock))

    def is_open(self):
        if self.sock is None:
            return False
        return not self.sock._closed

File: Lockard_PiComAPI/src/main/java/PiComAPI/LANClient.java
package PiComAPI;
import PiComAPI.Payload.Payload;
import PiComAPI.Payload.PayloadIntr;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.List;



public class LANClient {

    private String host;
    private int port;
    private Socket socket;

    public LANClient(String host, int port) {
        this.host = host;
        this.port = port;
        this.socket = new Socket();
        openConnection();
    }

    public void send(List<PayloadIntr> payloads, LANClientHandler handler){
        if (socket.isClosed())
            openConnection();

        for (PayloadIntr payload: payloads) {
            ComUtils.sendPayload(socket, payload);
            handler.received(payload, ComUtils.receivePayload(socket));
        }
    }


    public Payload send(Payload payload, LANClientHandler handler){
        if (socket.isClosed())
            openConnection();
        ComUtils.sendPayload(socket, payload);
        Payload p = ComUtils.receivePayload(socket);
        if (handler != null)
            handler.received(payload, p);
        return p;

    }

    public Payload send(Payload payload){
        return send(payload, null);
    }

    public void openConnection(){
        if (socket.isConnected())
            return;

        try {
            socket.setSoTimeout(ComUtils.TIMEOUT);
            socket.connect(new InetSocketAddress(host, port));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void closeConnection(){

        try {
            if (!socket.isClosed()){
                socket.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public boolean isOpen(){
        return socket.isConnected();
    }
}

